mmpy_bot/plugins/base.py

```diff
@@ -88,21 +70,213 @@ async def call_function(
+):
+    """Build FunctionInfo objects from plugin and function information.
+
+    Returns one FunctionInfo instance for every listener (message or webhook)
+    """
+
+    plug_help: List[FunctionInfo] = []
+
+    for matcher, functions in listeners.items():
+        for function in functions:
+            plug_head, plug_full = split_docstring(function.plugin.__doc__)
+            func_head, func_full = split_docstring(function.docstring)
+
+            if isinstance(function, MessageFunction):
+                direct = function.direct_only
+                mention = function.needs_mention
+                help_type = "message"
+            elif isinstance(function, WebHookFunction):
+                direct = mention = False
+                help_type = "webhook"
+            else:
+                raise NotImplementedError(
+                    f"Unknown/Unsupported listener type: '{type(function)}'"
+                )
+
+            plug_help.append(
+                FunctionInfo(
+                    help_type=help_type,
+                    location=function.plugin.__class__.__name__,
+                    function=function,
+                    pattern=matcher.pattern,
+                    plugin_docheader=plug_head,
+                    plugin_docfull=plug_full,
+                    function_docheader=func_head,
+                    function_docfull=func_full,
+                    direct=direct,
+                    mention=mention,
+                    is_click=function.is_click_function,
+                    metadata=function.metadata,
+                )
+            )
+
+    return plug_help
+
+
+class PluginManager:
+    """PluginManager is responsible for initializing all plugins and display aggregated
+    help from each of them.
+
+    It is supposed to be transparent to EventHandler that interacts directly with each
+    individual Plugin.
+    """
+
+    def __init__(
+        self,
+        plugins: Sequence[Plugin],
+    ):
+        self.settings: Optional[Settings] = None
+        self.plugins = plugins
+
+        self.message_listeners: Dict[re.Pattern, List[MessageFunction]] = defaultdict(
+            list
+        )
+        self.webhook_listeners: Dict[re.Pattern, List[WebHookFunction]] = defaultdict(
+            list
+        )
+
+    def initialize(self, driver: Driver, settings: Settings):
+        for plugin in self.plugins:
+            plugin.initialize(driver, self, settings)
+
+            # Register listeners for any listener functions in the plugin
+            for attribute in dir(plugin):
+                attribute = getattr(plugin, attribute)
+                if not isinstance(attribute, Function):
+                    continue
+
+                # Register this function and any potential siblings
+                for function in [attribute] + attribute.siblings:
+                    # Plugin message/webhook handlers can be decorated multiple times
```

Comment:
Why have you moved this functionality out of the `Plugin` class? I think accessing some `Plugin` attributes in a for loop here (generated by e.g. `Plugin.initialize()`) is a lot cleaner than this triple-nested for loop...